### PR TITLE
Fix marker overlap

### PR DESCRIPTION
**Pull request by jakeprojects**
## PR Summary

Performance optimization of sorting by z co-ordinates for scatter plots of Path3DCollection objects. Locally observed a 20x speed-up using this approach, unit tests test_mplot3d.py passed. Initially forked from xenoryt.

Reuses #10791 with performance optimization.

## PR Checklist

- [x] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [x] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/mpl_toolkits/mplot3d/art3d.py [hunk 1]
```diff
@@ -544,9 +544,27 @@ def do_3d_projection(self, renderer):
...
         ecs = (_zalpha(self._edgecolor3d, vzs) if self._depthshade else
                self._edgecolor3d)
+
+        # Sort the points based on z coordinates
+        # Performance optimization: Create a sorted index array and reorder
+        # points and point properties according to the index array
+        z_markers_idx = np.argsort(vzs)[::-1]
+
+        # Re-order items
+        zzs = vzs[z_markers_idx]
```
Comment (tacaswell) [inline lib/mpl_toolkits/mplot3d/art3d.py]: Was this rename intentional

File: lib/mpl_toolkits/mplot3d/art3d.py [hunk 2]
```diff
@@ -691,24 +709,23 @@ def do_3d_projection(self, renderer):
             else:
                 cedge = cedge.repeat(len(xyzlist), axis=0)
 
-        # sort by depth (furthest drawn first)
+        # # sort by depth (furthest drawn first)
```
Comment (anntzer) [inline lib/mpl_toolkits/mplot3d/art3d.py]: undo this change

File: lib/mpl_toolkits/mplot3d/art3d.py [hunk 3]
```diff
@@ -691,24 +709,23 @@ def do_3d_projection(self, renderer):
...
         z_segments_2d = sorted(
             ((self._zsortfunc(zs), np.column_stack([xs, ys]), fc, ec, idx)
              for idx, ((xs, ys, zs), fc, ec)
              in enumerate(zip(xyzlist, cface, cedge))),
             key=lambda x: x[0], reverse=True)
 
-        segments_2d = [s for z, s, fc, ec, idx in z_segments_2d]
+        [zzs, segments_2d, self._facecolors2d, self._edgecolors2d, idxs] = \
```
Comment (anntzer) [inline lib/mpl_toolkits/mplot3d/art3d.py]: the brackets are unnecessary

**Comment by QuLogic** [comment]
@jakeprojects are you still interested in finishing this?

**Comment by timhoffm** [comment]
I took the liberty to fix the minor comments above and rebase on master.

**Comment by QuLogic** [comment]
Not passing though...

**Comment by timhoffm** [comment]
Tests were broken because the test and reference images appeared to have a different magnification (or layout margin). Not sure if we changed anything in this respect.

Anyway, I've exchanged the image comparisons with figure comparison tests checking:

1. The data point order does not influence which marker is drawn on top.
2. Rotating the view by 180° and exchanging the colors should give the original image.

**Commits should be squashed before merging to remove the test images from the earlier commits.** I didn't want to force push a squash on another user's repo and remove part of the test code without approval.

**Comment by QuLogic** [comment]
I'd rather they be rebased out instead of squashing, because the commits are written by different authors since this was taken up from a previous PR. I have not checked yet if the new test is okay though.

**Comment by timhoffm** [comment]
Not sure how that rebase should work. The first commit contains the image tests and part of the relevant code. Do you wish to strip the tests from that commit?

I think the best is to squash everything into one commit because we do not need the back-and-forth of the development history here. We can credit all the authors for the commit.

File: lib/mpl_toolkits/mplot3d/art3d.py [hunk 4]
```diff
@@ -494,9 +494,27 @@ def do_3d_projection(self, renderer):
...
+        # points and point properties according to the index array
+        z_markers_idx = np.argsort(vzs)[::-1]
+
+        # Re-order items
+        vzs = vzs[z_markers_idx]
+        vxs = vxs[z_markers_idx]
+        vys = vys[z_markers_idx]
+        fcs = fcs[z_markers_idx]
+        ecs = ecs[z_markers_idx]
+        vps = np.vstack((vxs, vys)).T
```
Comment (QuLogic) [inline lib/mpl_toolkits/mplot3d/art3d.py:509]: Why change to `vstack.T` instead of `column_stack`?

Comment (timhoffm) [reply lib/mpl_toolkits/mplot3d/art3d.py:509]: Both are fine. I've changed to `column_stack`, which seems a bit clearer.

Interestingly, `vstack.T` is marginally faster for > 10k elements:
grafik
But 10ms difference is not the end of the world also for 1e6 elements. Probably all the surrounding code is much slower anyway.

**Comment by timhoffm** [comment]
#17378: Same code but commits squashed and images stripped. (I didn't want to push the squash to another user's repo, so I had to open a separate PR. All three contributors are credited.

**Comment by QuLogic** [comment]
Closing in favour of the above squashed PR.
